src/callbacks.cpp

```diff
@@ -0,0 +1,201 @@
+        if (msg.active) {
+            this->activate();
+        } else {
+            this->deactivate();
+        }
+    }
+
+    // If the node should be deactivated, don't parse the payload as it is only
+    // relevant in active mode
+    if (!this->get_active()) {
+        return;
+    }
+
+    nlohmann::json cmd_json;
+    try {
+        cmd_json = common_lib::CommandDefinitions::parse_check_json_str(
+            msg.payload,
+            common_lib::CommandDefinitions::get_waypoint_command_definition());
+    } catch (const std::runtime_error &e) {
+        RCLCPP_FATAL(this->get_logger(),
+                     "WaypointNode::callback_control: Received invalid json as "
+                     "payload: %s",
+                     e.what());
+        this->job_finished((std::string) "WaypointNode::callback_control: Received "
+                                         "invalid json as payload: " +
+                           e.what());
+        reset_node();
+        return;
+    }
+
+    // Set flag to true to indicate that the object has real values in it
+    cmd.values_set = true;
+
+    // Required parameters
+    cmd.target_coordinate_lat = cmd_json.at("target_coordinate_lat");
+    cmd.target_coordinate_lon = cmd_json.at("target_coordinate_lon");
+    cmd.cruise_height_cm = cmd_json.at("cruise_height_cm");
+    cmd.target_height_cm = cmd_json.at("target_height_cm");
+    cmd.horizontal_speed_mps = cmd_json.at("horizontal_speed_mps");
+    cmd.vertical_speed_mps = cmd_json.at("vertical_speed_mps");
+
+    // Optional parameters
+    if (cmd_json.contains("pre_wait_time_ms"))
+        cmd.pre_wait_time_ms = cmd_json.at("pre_wait_time_ms");
+
+    if (cmd_json.contains("post_wait_time_ms"))
+        cmd.post_wait_time_ms = cmd_json.at("post_wait_time_ms");
+}
+
+/**
+ * @brief Callback function for receiving GPS position messages.
+ *
+ * This function is called when a GPS position message is received. It extracts
+ * the latitude, longitude, and height information from the message and updates
+ * the corresponding variables in the `pos` object.
+ *
+ * @param msg The GPS position message containing the latitude, longitude, and
+ * height information.
+ */
+void WaypointNode::callback_position(const interfaces::msg::GPSPosition &msg) {
+    RCLCPP_DEBUG(this->get_logger(),
+                 "WaypointNode::callback_position: Received position from "
+                 "'%s': lat: %f, lon: %f, height: %f",
+                 msg.sender_id.c_str(), msg.latitude_deg, msg.longitude_deg,
+                 msg.relative_altitude_m);
+
+    // Check timestamp
+    rclcpp::Time timestamp_now = this->now();
+    if (timestamp_now - rclcpp::Time(msg.time_stamp) >
+        rclcpp::Duration(std::chrono::duration<int64_t, std::milli>(
+            max_position_msg_time_difference_ms))) {
+        // Reset pos as we can no longer reliabely know where we are
+        pos = Position();
+
+        if (this->get_active()) {
```

Comment:
Add comment above this paragraph that states that this path aborts the program as the received pos was too old

Reply:
Fixed in https://github.com/THI-Drone/waypoint_package/pull/3/commits/5a49068ac86cdab12bbed671d77177ca7fee6d9a